packages/cw721/src/msg.rs

```diff
@@ -21,15 +50,640 @@ pub enum Cw721ExecuteMsg {
         expires: Option<Expiration>,
     },
     /// Remove previously granted Approval
-    Revoke { spender: String, token_id: String },
+    Revoke {
+        spender: String,
+        token_id: String,
+    },
     /// Allows operator to transfer / send any token from the owner's account.
     /// If expiration is set, then this allowance has a time/height limit
     ApproveAll {
         operator: String,
         expires: Option<Expiration>,
     },
     /// Remove previously granted ApproveAll permission
-    RevokeAll { operator: String },
+    RevokeAll {
+        operator: String,
+    },
+
+    /// Mint a new NFT, can only be called by the contract minter
+    Mint {
+        /// Unique ID of the NFT
+        token_id: String,
+        /// The owner of the newly minter NFT
+        owner: String,
+        /// Universal resource identifier for this NFT
+        /// Should point to a JSON file that conforms to the ERC721
+        /// Metadata JSON Schema
+        token_uri: Option<String>,
+        /// Any custom extension used by this contract
+        extension: TNftMetadataExtensionMsg,
+    },
+
     /// Burn an NFT the sender has access to
-    Burn { token_id: String },
+    Burn {
+        token_id: String,
+    },
+
+    /// Metadata msg
+    #[deprecated(since = "0.19.0", note = "Please use UpdateNftMetadata instead")]
+    /// Deprecated: use UpdateNftMetadata instead! In previous release it was a no-op for customization in other contracts. Will be removed in next release!
+    Extension {
+        msg: TNftMetadataExtensionMsg,
+    },
```

Comment:
Just to be clear, this `ExecuteExtension` was previously about more than updating metadata, you could additional messages to execute. For example: https://github.com/CosmosContracts/code-and-chill/blob/main/contracts/cw721-piggy-bank/src/contract.rs#L83

I'm fine with a different pattern (personally think this is all much better in Sylvia), also not sure how many people have built custom extensions adding new methods or queries to their NFTs. How do you propose people add custom Execute and Query messages. For example, say there are some monster NFTs that you can feed with crypto for them to get bigger, how can we support people building such a thing?

Reply:
excellent point! before PR there was a contradictional design. generic was named `T` (now `TNftMetadataExtensionMsg`) in both query and execute traits:
a) in execute it (may) be used for storing extension
b) in query it (may) be used for custom queries

Using a) or b) is fine. Previously Cw721Contract implemented both traits by using same generic `T` for both traits. That's wrong because by using both cases would be contradictionary. PR reveals this.

So for execute trait it is sufficient providing an incoming message for T.
For query it needs to provide incoming and outgoing msg. Something like `TNftMetadataExtensionQueryMsg` and `TNftMetadataExtensionResponse`. Will go over it. But this means there will be another generic - increasing from 5 to 6 generics. But fine imo.

Reply:
@JakeHartnell, alright. Design flaw is in usage of `Q` here (taken before my changes):

```rs
pub enum QueryMsg<Q: JsonSchema> {
...
    #[returns(cw721::NftInfoResponse<Q>)] // here Q is then extension/onchain metadata
    NftInfo { token_id: String },
...
    #[returns(())]
    Extension { msg: Q }, // here Q is a custom query msg
...
```

If I replace Q in Extension to `Extension { msg: TExtensionQueryMsg }` then I get this compile error, since `Q` cannot be inferred:

> cannot infer type for type parameter `TNftExtension` (formerly `Q`) declared on the enum `Cw721QueryMsg` (formerly QueryMsg)

So in your code example, defining `Q` as `QueryExt` also affects `NftInfo` query, where it expects extension/onchain metadata of type `QueryExt`.

So this can only be solved by defining 2 msgs, where onchain data and extension msg is part of it. So I will add these 2:

- `NftExtension{filter: Q}`: this will be a new query where it finds first NFT with given traits
- `Extension { msg: TExtensionQueryMsg }`: for custom extension query

Once done, I'll look into execute msg.

Reply:
@JakeHartnell, fixed here: https://github.com/CosmWasm/cw-nfts/pull/156#discussion_r1536914085
And here: https://github.com/CosmWasm/cw-nfts/pull/156#discussion_r1536916211